File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Scenarios/RcB3MessagePackCodecScenario.java
package systems.zlink.e2e.registrationcodec.client.Scenarios;

import systems.zlink.e2e.registrationcodec.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioContext;
import systems.zlink.e2e.registrationcodec.shared.Contracts;

public final class RcB3MessagePackCodecScenario {
    private RcB3MessagePackCodecScenario() {
    }

    public static void run(ScenarioContext context) {
        String packedValue = context.server()
            .post("/codec/roundtrip")
            .submit(Contracts.CodecScenarioRes.class).toCompletableFuture().join().body()
            .messagePackValue();
        ScenarioAssert.ensure("echo:msgpack-request".equals(packedValue),
            "RC-B3 request mismatch");
        ScenarioAssert.waitForEvidence(
            context.evidence(),
            "ContentType",
            "MsgpackEcho",
            "application/x-msgpack");
        ScenarioAssert.waitForEvidence(context.evidence(), "Send", "MsgpackEcho", "msgpack-send");
        System.out.println("scenario RC-B3 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/client/Scenarios/RcB3MessagePackCodecScenario.kt
package systems.zlink.e2e.kotlin.registrationcodec.client.scenarios

import systems.zlink.e2e.kotlin.registrationcodec.PackedEchoRes
import systems.zlink.e2e.kotlin.registrationcodec.client.support.CodecScenarioResult
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioAssert
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioHttpClient

class RcB3MessagePackCodecScenario(
    private val server: ScenarioHttpClient,
    private val assert: ScenarioAssert,
) {
    fun run(): CodecScenarioResult {
        val packedReply = server.post<PackedEchoRes>("/codec/messagepack")
        assert.that(packedReply.value == "echo:msgpack-request", "RC-B3 request mismatch")
        assert.that(
            server.waitForEvidence("Send", "PackedEchoMsg", "msgpack-send")
                .any { it.marker == "Send" && it.packetName == "PackedEchoMsg" && it.value == "msgpack-send" },
            "RC-B3 send evidence missing",
        )
        println("scenario RC-B3 passed")
        return CodecScenarioResult("RC-B3", "PackedEchoReq", packedReply.value)
    }
}
